### PR TITLE
Chainmail: Rewrite in Rust

### PR DESCRIPTION
**Pull request by cool-ant**
- rewrite service in Rust

### CODE REVIEW

**Review by James-Mart — DISMISSED** [review]
You should delete the `Webmail/src/Webmail.cpp` and `Webmail/CMakeLists.txt` files which were only used when the service was implemented in C++. The header file you could leave for the C++ wrapper.

File: services/user/Webmail/service/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+    use crate::{serve_rest_api, validate_user};
+
+    #[table(record = "WebContentRow")]
+    struct WebContentTable;
+
+    #[action]
+    fn send(receiver: AccountNumber, subject: String, body: String) {
+        check(
+            validate_user(receiver.to_string()),
```
[the remark this answers is not in the document]
Comment (cool-ant) [reply services/user/Webmail/service/src/lib.rs:117]: The functionality provided by validate_user() is used in 3 different places, so I think it's useful to DRY things up a bit (the idea is mirroring your decision in your original code as well), but point taken about extra steps; I've rewritten validate_user(...).

File: services/user/Webmail/service/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+            where_clause += r_clause.as_str();
+        }
+
+        let mq = make_query(
+            request,
+            format!(
+                "SELECT * FROM \"history.webmail.sent\" {} ORDER BY ROWID",
+                where_clause
+            )
+            .as_str(),
```
Comment (swatanabe) [inline services/user/Webmail/service/src/lib.rs:90]: `make_query` can just take a `String` and avoid going from `String` -> `&str` -> `String`

File: rust/psibase/src/services/events.rs [hunk 3]
```diff
@@ -0,0 +1,15 @@
+#[crate::service(name = "events", dispatch = false, psibase_mod = "crate")]
+#[allow(non_snake_case, unused_variables)]
+mod service {
+    fn setSchema(schema: &crate::Schema) {
```
Comment (swatanabe) [inline rust/psibase/src/services/events.rs:4]: Missing `#[action]`

File: services/user/Webmail/service/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        }
+
+        let query_start = request.target.find('?');
+        if query_start.is_none() {
+            return None;
+        }
+        let query_start = query_start.unwrap();
+
+        let query = request.target.split_at(query_start + 1).1;
+        let params = crate::parse_query(String::from(query));
```
[the remark this answers is not in the document]
Comment (cool-ant) [reply services/user/Webmail/service/src/lib.rs:53]: ah, good catch

File: services/user/Webmail/service/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+        let mut s_clause = String::new();
+        let s_opt = params.get(&String::from("sender"));
+        if let Some(s) = s_opt {
+            validate_user(AccountNumber::from(s.as_str()));
```
[the remark this answers is not in the document]
Comment (cool-ant) [reply services/user/Webmail/service/src/lib.rs:57]: All rest calls are failing due to the rewrite. James has a PR up (#847 ) to attempt a postinstall to init the index the rest calls rely on. The 500s are coming from the event index not being built.

[the remark this answers is not in the document]
Comment (swatanabe) [reply services/user/Webmail/service/src/lib.rs:57]: That's a completely different matter. Returning `std::nullopt`/`None` results in a 404. Aborting results in 500.

**Review by swatanabe — APPROVED** [review]

**Pull request merged**